kernels/histogram.cl

```diff
@@ -0,0 +1,40 @@
+// adapted code from https://github.com/bgaster/opencl-book-samples/blob/master/src/Chapter_14/histogram/histogram_image.cl
+
+#pragma OPENCL EXTENSION cl_khr_local_int32_base_atomics : enable
+
+const sampler_t sampler = CLK_NORMALIZED_COORDS_FALSE | CLK_ADDRESS_CLAMP_TO_EDGE | CLK_FILTER_NEAREST;
+
+kernel void histogram(
+    IMAGE_src_TYPE src,
+    IMAGE_dst_TYPE dst,
+    float minimum,
+    float maximum,
+    int step_size_x,
+    int step_size_y,
+    int step_size_z
+)
+{
+    const int image_width  = GET_IMAGE_WIDTH(src);
+    const int image_height = GET_IMAGE_HEIGHT(src);
+    const int image_depth  = GET_IMAGE_DEPTH(src);
+    
+    const int y = get_global_id(0) * step_size_y;
+    float range = maximum - minimum;
+
+    uint tmp_histogram[NUMBER_OF_HISTOGRAM_BINS];
+    for (int i = 0; i < NUMBER_OF_HISTOGRAM_BINS; ++i) {
+        tmp_histogram[i] = 0;
+    }
+
+    for (int z = 0; z < GET_IMAGE_DEPTH(src); z += step_size_z) {
+        for (int x = 0; x < GET_IMAGE_WIDTH(src); x += step_size_x) {
+            float clr = READ_IMAGE(src, sampler, POS_src_INSTANCE(x,y,z,0)).x;
+            uint indx_x = convert_uint_sat( ( (clr - minimum) * (GET_IMAGE_WIDTH(dst) - 1) ) / range + 0.5);
```

Comment:
I am facing an issue here where, in 1D only apparently, i get a shift in value during computation of the histogram, most likely from a rounding issue.

```
input  : 7 0 3 9 4 4 2 2 4 9 
valid  : 1 0 2 1 3 0 0 1 0 2 
output : 1 0 2 1 3 0 0 1 0 2 // all went well
```

```
input  : 1 6 9 4 6 9 5 5 7 8 
valid  : 0 1 0 0 1 2 2 1 1 2 
output : 1 0 0 1 0 2 2 1 1 2  // rounding issue for unknown reason in the middle of the process
```

no clear reason to where this come from

Reply:
Ok so no, the reason is linked to the absence of 0, hence `min_max=(1,9)` which leads to index 0 when value is 1.
The kernel only works when the number of bins is the length of the intensity range. Is this fine?